posts/2020-02-12-using-dotty-union-types-with-akka-typed-part-II.adoc

```diff
@@ -0,0 +1,285 @@
+
+== Restoring the `Pinger` actor's protocol
+
+Let's have a look at the final version of the protocol of the `Pinger` actor using Scala 2.13 and Akka 2.6:
+
+[source,scala,linenums]
+----
+object Pinger {
+
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+  final case class WrappedPongResponse(pong: PingPong.Response)
+    extends Command
+
+  ...
+}
+----
+
+The problem should be obvious: we shouldn't expose the `WrappedPongResponse` wrapper message as part of the `Pinger` actor's protocol. In other words, we have unnecessarily _widened_ the actor's protocol by adding the publicly accessible `WrappedPongResponse` message to it.
+
+Obviously, this is undesirable and the fix in this case is quite simple: we can mark this message as private and we're done:
+
+[source,scala,linenums]
+----
+object Pinger {
+
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+  private final case class WrappedPongResponse(pong: PingPong.Response)
+    extends Command
+
+  ...
+}
+----
+
+== Restoring the `Pinger` actor's protocol - Dotty version
+
+The Dotty version of the `Pinger` actor suffers from the same problem. The protocol of the actor is _widened_ with the `PingPong.Pong` message. This means that _any_ actor can send a `Pong` message to the `Pinger` actor. The solution in this case is different from previous one:
+
+* The `apply` method in the `Pinger` object, will return a `Behavior[Command]` thus restoring the original protocol
+* The `Pinger` actor defines an internal behavior of type `Behavior[CommandIncludingResponses]`
+* The `Behavior[Command]` is derived from the `Behavior[CommandIncludingResponses]` by applying the `narrow` method on the latter
+
+This leads to the following code:
+
+[source,scala,linenums]
+----
+object Pinger {
+
+  // My protocol
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+
+  // My protocol + the responses I need to understand...
+  private type CommandsAndResponses = Command | PingPong.Response
+
+  def apply(pingPong: ActorRef[PingPong.Ping]): Behavior[Command] = {
+    val internalBehavior: Behavior[CommandsAndResponses] =
+    Behaviors.setup { context =>
+      Behaviors.receiveMessage {
+        case StopPingPong =>
+          context.log.info(s"End of the ping-pong game - calling it a day!")
+          context.system.terminate()
+          Behaviors.stopped
+        case SendPing =>
+          pingPong ! PingPong.Ping(replyTo = context.self)
+          Behaviors.same
+        case response : PingPong.Response =>
+          context.log.info(s"Hey: I just received a $response !!!")
+          Behaviors.same
+      }
+    }
+    internalBehavior.narrow
+  }
+}
+----
+
+Admittedly, some magic has happened here and the question to ask is, _What is this `narrow` method doing?_. Let's look at that in the next section.
```

Comment:
Just and extra period `.`, after the question mark

Reply:
Good catch. Removed the '.' and moving the question mark outside the italic formatting.